File: ainpc-core-plugin/src/main/java/ro/ainpc/engine/runtime/ScenarioExecutionContext.java
package ro.ainpc.engine.runtime;

import java.util.LinkedHashMap;
import java.util.Map;

public record ScenarioExecutionContext(
    String playerUuid,
    String playerName,
    String npcId,
    String npcName,
    String regionId,
    String placeId,
    String nodeId,
    String templateId,
    String progressionId,
    String runtimeMode,
    Map<String, String> variables
) {
    public ScenarioExecutionContext {
        playerUuid = valueOrEmpty(playerUuid);
        playerName = valueOrEmpty(playerName);
        npcId = valueOrEmpty(npcId);
        npcName = valueOrEmpty(npcName);
        regionId = valueOrEmpty(regionId);
        placeId = valueOrEmpty(placeId);
        nodeId = valueOrEmpty(nodeId);
        templateId = valueOrEmpty(templateId);
        progressionId = valueOrEmpty(progressionId);
        runtimeMode = valueOrEmpty(runtimeMode);
        variables = sanitize(variables);
    }

    public String variable(String key) {
        return variables.getOrDefault(key, "");
    }

    private static String valueOrEmpty(String value) {
        return value == null ? "" : value.trim();
    }

    private static Map<String, String> sanitize(Map<String, String> source) {
        if (source == null || source.isEmpty()) {
            return Map.of();
        }
        Map<String, String> sanitized = new LinkedHashMap<>();
        source.forEach((key, value) -> {
            String safeKey = valueOrEmpty(key);
            if (!safeKey.isBlank()) {
                sanitized.put(safeKey, valueOrEmpty(value));
            }
        });
        return Map.copyOf(sanitized);
    }
}

File: ainpc-core-plugin/src/main/kotlin/ro/ainpc/engine/runtime/ScenarioExecutionContext.kt
package ro.ainpc.engine.runtime

import java.util.LinkedHashMap

class ScenarioExecutionContext(
    playerUuid: String?,
    playerName: String?,
    npcId: String?,
    npcName: String?,
    regionId: String?,
    placeId: String?,
    nodeId: String?,
    templateId: String?,
    progressionId: String?,
    runtimeMode: String?,
    variables: Map<String, String>?
) {
    private val playerUuidValue: String = valueOrEmpty(playerUuid)
    private val playerNameValue: String = valueOrEmpty(playerName)
    private val npcIdValue: String = valueOrEmpty(npcId)
    private val npcNameValue: String = valueOrEmpty(npcName)
    private val regionIdValue: String = valueOrEmpty(regionId)
    private val placeIdValue: String = valueOrEmpty(placeId)
    private val nodeIdValue: String = valueOrEmpty(nodeId)
    private val templateIdValue: String = valueOrEmpty(templateId)
    private val progressionIdValue: String = valueOrEmpty(progressionId)
    private val runtimeModeValue: String = valueOrEmpty(runtimeMode)
    private val variablesValue: Map<String, String> = sanitize(variables)

    fun playerUuid(): String = playerUuidValue
    fun playerName(): String = playerNameValue
    fun npcId(): String = npcIdValue
    fun npcName(): String = npcNameValue
    fun regionId(): String = regionIdValue
    fun placeId(): String = placeIdValue
    fun nodeId(): String = nodeIdValue
    fun templateId(): String = templateIdValue
    fun progressionId(): String = progressionIdValue
    fun runtimeMode(): String = runtimeModeValue
    fun variables(): Map<String, String> = variablesValue

    fun variable(key: String?): String = variablesValue[key].orEmpty()

    companion object {
        private fun valueOrEmpty(value: String?): String = value?.trim().orEmpty()

        private fun sanitize(source: Map<String, String>?): Map<String, String> {
            if (source.isNullOrEmpty()) {
                return emptyMap()
            }
            val sanitized = LinkedHashMap<String, String>()
            source.forEach { (key, value) ->
                val safeKey = valueOrEmpty(key)
                if (safeKey.isNotBlank()) {
                    sanitized[safeKey] = valueOrEmpty(value)
                }
            }
            return java.util.Map.copyOf(sanitized)
        }
    }
}
